libs/datasets/combined_datasets.py

```diff
@@ -269,28 +254,126 @@ def _build_combined_dataset_from_sources(
+    datasource_dataframes: Mapping[str, pd.DataFrame],
+    override=Override.BY_ROW,
+) -> pd.DataFrame:
     # structlog makes it very easy to bind extra attributes to `log` as it is passed down the stack.
     log = structlog.get_logger()
-    for field, data_source_classes in feature_definition_config.items():
-        for data_source_cls in data_source_classes:
-            dataset = intermediate_datasets[data_source_cls]
-            with tmp_bind(log, dataset_name=data_source_cls.SOURCE_NAME, field=field) as log:
-                try:
-                    data = dataset_utils.fill_fields_with_data_source(
-                        log, data, dataset.data, target_dataset_cls.INDEX_FIELDS, [field]
-                    )
-                except Exception:
-                    log.exception("trying to fill fields")
-                    raise
 
-    return target_dataset_cls(data)
+    # These are columns that are expected to have a single value for each FIPS. Get the columns
+    # from every row of each data source and then keep one of each unique row.
+    preserve_columns = [
+        CommonFields.AGGREGATE_LEVEL,
+        CommonFields.STATE,
+        CommonFields.COUNTRY,
+        CommonFields.COUNTY,
+    ]
+    all_identifiers = pd.concat(
+        df.reset_index().loc[
+            :, [CommonFields.FIPS] + list(df.columns.intersection(preserve_columns))
+        ]
+        for df in datasource_dataframes.values()
+    ).drop_duplicates()
+    # Make a DataFrame with a unique FIPS index. If multiple rows are found with the same FIPS then there
+    # are rows in the input data sources that have different values for county name, state etc.
+    fips_indexed = all_identifiers.set_index(CommonFields.FIPS, verify_integrity=True)
+
+    # Inspired by pd.Series.combine_first(). Create a new index which is a union of all the input dataframe
+    # index.
+    dataframes = list(datasource_dataframes.values())
+    new_index = dataframes[0].index
+    for df in dataframes[1:]:
+        new_index = new_index.union(df.index)
+    # Override.BY_ROW needs to preserve the rows of the input dataframes. If not going BY_ROW
+    # reindex the inputs now to avoid reindexing for each field below.
+    if override is not Override.BY_ROW:
+        datasource_dataframes = {
+            name: df.reindex(new_index, copy=False) for name, df in datasource_dataframes.items()
+        }
+
+    # Build feature columns from feature_definitions.
+    data = pd.DataFrame(index=new_index)
+    for field_name, data_source_names in feature_definitions.items():
+        log.info("Working field", field=field_name)
+        field_out = None
+        # Go through the data sources, starting with the highest priority.
+        for datasource_name in reversed(data_source_names):
+            with tmp_bind(log, dataset_name=datasource_name, field=field_name) as log:
+                datasource_field_in = datasource_dataframes[datasource_name][field_name]
```

Comment:
Thoughts on pulling this series replacement code out into its own function? I don't think it's necessary, but would make test writing a bit easier for this (fairly complicated) logic.

Reply:
Proof by example that testing the complex logic is possible as-is... at least I'm pretty sure  test_build_timeseries_override exercises each of the override types.